fetch/fetch.go

```diff
@@ -0,0 +1,169 @@
+// should use this client.
+var client *http.Client
+
+// Fetcher provides a interface for anything that can get RSS data and provide it in a
+// sequential fashion (e.g. without concurrency). The fetcher is the building block for
+// larger subscription routines that periodically use the fetcher to retrieve data.
+// Fetchers should therefore be treated as things that will only run inside of a single
+// thread, whereas Subscription objects are things that may run concurrently.
+type Fetcher interface {
+	Fetch() (feed *gofeed.Feed, err error)
+}
+
+// New creates a new HTTP fetcher that can fetch rss feeds from the specified URL.
+func New(url string) Fetcher {
+	return &httpFetcher{
+		url:    url,
+		parser: gofeed.NewParser(),
+	}
+}
+
+// SetDefaultClient allows you to specify an alternative http.Client to the default one
+// used by all http based Fetchers in this package. Use this function to change the
+// timeouts of the client or to set a test client.
+func SetDefaultClient(c *http.Client) {
+	client = c
+}
+
+//===========================================================================
+// HTTP Fetcher
+//===========================================================================
+
+// The httpFetcher uses GET requests to retrieve data with a Baleen-specific http
+// client. We avoid using gofeed.ParseURL because it is very simple and doesn't respect
+// rate limits or etags, which are necessary for Baleen to run in continuous operation.
+type httpFetcher struct {
+	url      string         // the url of the RSS or atom feed
+	parser   *gofeed.Parser // the universal feed parser for RSS and Atom feeds
+	etag     string         // used for conditional http to minimize bandwidth
+	modified string         // used for conditional http to minimize bandwidth
+}
+
+func (f *httpFetcher) Fetch() (feed *gofeed.Feed, err error) {
+	var req *http.Request
+	if req, err = f.newRequest(); err != nil {
+		return nil, err
+	}
+
+	var rep *http.Response
+	if rep, err = client.Do(req); err != nil {
+		return nil, err
+	}
+
+	// Close the body of the response reader when we're done.
+	if rep != nil {
+		defer func() {
+			ce := rep.Body.Close()
+			if ce != nil {
+				err = ce
+			}
+		}()
+	}
+
+	// Check the status code of the response; note that 304 means not modified, but we
+	// are still returning a 304 error to signal to the Subscription that nothing has
+	// changed and that the feed is nil.
+	if rep.StatusCode < 200 || rep.StatusCode >= 300 {
+		return nil, HTTPError{
+			Status: rep.Status,
+			Code:   rep.StatusCode,
+		}
+	}
+
+	// Use the universal parser to parse the Atom or RSS feed
+	if feed, err = f.parser.Parse(rep.Body); err != nil {
+		return nil, err
+	}
+
+	// Get the eTag and last-modified from the response header if we've successfully
+	// parsed the request and received a 200 response.
+	f.etag = rep.Header.Get("ETag")
+	f.modified = rep.Header.Get("Last-Modified")
```

Comment:
Ok, after looking at this more closely and thinking it through, I *think* 126 and 127 are in the right place; let me know if my reasoning makes sense:

If `Fetch` returns a feed and no error, the state of the fetcher should have also been updated. If `Fetch` returns an error that is either a panic (which is [what looks like what `NewRequest` will do on err](https://golang.org/pkg/net/http/httptest/#NewRequest)), a `url.Error` (from `client.Do(req)`), an `HTTPError`, or a parsing error, the feed will be nil and the fetcher's state won't change. 

Like you said, one possible problem with this is that that doesn't give us an easy way to tell how long a fetcher has had the same state relative to the number of times it has attempted to `Fetch`, which could (maybe?) mask a long running series of errors in a very long running fetcher. Would it be a bad idea to give each fetcher a monotonically increasing attribute that would allow us to diagnose with heuristics a case where a fetcher has had `Fetch` called unsuccessfully many times in a row? Or maybe this would just be easier to deal with on the user/app side?

Reply:
I think a monotonically increasing counter is a good idea, and in CH240 I discuss adding something like these counters to the database. My gut says if we apply heuristics to figure out what's happening with a url; we should do it on the Subscription side to keep the fetcher as simple as possible. I'll add an issue to the backlog for us to think about it further!